### PR TITLE
Custom metric to track stalled attempts

### LINKED ISSUE #1559: Poll the number of Attempts that remain in an available state past a configured time threshold
Add a custom polling metric to produce a count of Attempt records that have been in the `available` state longer than a given threshold.

### PR DESCRIPTION
**Pull request by rorymckinley**
## Notes for the reviewer

Adds a custom metric to track stalled attempts. I had to add to some magic with `:sys.get_state` to ensure that PromEx does not try to execute the metric before `Repo` is ready.

## Related issue

Fixes #1559 

## Review checklist

- [x] I have performed a **self-review** of my code
- [x] I have verified that all appropriate **authorization policies** have been implemented and tested
- [x] If needed, I have updated the **changelog**
- [ ] Product has **QA'd** this feature

### CODE REVIEW

File: lib/lightning/attempts/prom_ex_plugin.ex [hunk 1]
```diff
@@ -20,4 +24,60 @@ defmodule Lightning.Attempts.PromExPlugin do
...
+    query =
+      from a in Attempt,
+        select: count(a.id),
+        where: a.state == :available,
+        where: a.inserted_at < ^threshold_time
+
+    count = Repo.one(query)
+
+    :telemetry.execute(
+      [:lightning, :attempt, :queue, :stalled],
```
Comment (jyeshe) [inline lib/lightning/attempts/prom_ex_plugin.ex:78]: minor detail: to put the event in a constant (module attribute like `@event_name`) to make it clear that the one dispatched is the same of the publishing.

File: test/lightning/attempts/prom_ex_plugin_test.exs [hunk 2]
```diff
@@ -33,4 +36,114 @@ defmodule Lightning.Attempts.PromExPluginText do
...
+        [stalled_attempt_threshold_seconds]
+      }
+
+    polling = PromExPlugin.polling_metrics(%{})
+
+    assert polling.group_name == :lightning_attempt_polling_events
+
+    assert polling.poll_rate == 5000
+
+    assert polling.measurements_mfa == expected_mfa
```
Comment (jyeshe) [inline test/lightning/attempts/prom_ex_plugin_test.exs:59]: this is a suggested style to indicate that all assertions are related to same data:
```
assert %{
  group_name: :lightning_attempt_polling_events, 
  poll_rate: 5000, 
  measurements_mfa: ^expected_mfa, 
  metrics: [metric | _]
 } = PromExPlugin.polling_metrics(%{})
```

File: test/lightning/attempts/prom_ex_plugin_test.exs [hunk 3]
```diff
@@ -33,4 +36,114 @@ defmodule Lightning.Attempts.PromExPluginText do
...
+
+    [metric | _] = polling.metrics
+
+    assert %Telemetry.Metrics.LastValue{} = metric
```
Comment (jyeshe) [inline test/lightning/attempts/prom_ex_plugin_test.exs:63]: Would do the same here even though it is a small code. It's one less check for the brain to know if the assertions have something in common or not.

File: test/lightning/attempts/prom_ex_plugin_test.exs [hunk 4]
```diff
@@ -33,4 +36,114 @@ defmodule Lightning.Attempts.PromExPluginText do
...
+      now = DateTime.utc_now()
+      _stalled_1 = available_attempt(now, -60)
+      _stalled_2 = available_attempt(now, -50)
+      _stalled_2 = available_attempt(now, -40)
+      _not_stalled_due_to_time = available_attempt(now, -20)
+      _not_stalled_due_to_state = claimed_attempt(now, -40)
+
+      PromExPlugin.stalled_attempt_count(stall_threshold_seconds)
+
+      assert_received {
```
Comment (jyeshe) [inline test/lightning/attempts/prom_ex_plugin_test.exs:93]: :pray:

**Review by jyeshe — COMMENTED** [review]
Clear code with nice test coverage! Only few refactoring and style suggestions.

**Comment by rorymckinley** [comment]
Thanks @jyeshe . I have pushed the changes requested in  new commit - I will collapse it once you are happy that I addressed all your concerns.

**Review by jyeshe — APPROVED** [review]

**Pull request merged**